src/tt_scheduler.erl

```diff
@@ -0,0 +1,137 @@
+%%%-------------------------------------------------------------------
+%%% @author Wojciech Pachuta
+%%%-------------------------------------------------------------------
+-module(tt_scheduler).
+-author("Wojciech Pachuta").
+
+-behaviour(gen_server).
+
+%% API
+-export([start_link/0, schedule/3, cancel_schedule/1]).
+
+%% gen_server callbacks
+-export([init/1,
+  handle_call/3,
+  handle_cast/2,
+  handle_info/2,
+  terminate/2,
+  code_change/3]).
+
+%% auxiliary functions
+-export([time_keeper_function/4,  prompt_publishing/1]).
+
+-define(SERVER, ?MODULE).
+
+-record(state, {refs :: #{reference() => pid()}}).
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+-spec(start_link() ->
+  {ok, Pid :: pid()} | ignore | {error, Reason :: term()}).
+start_link() ->
+  gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+-spec(schedule(calendar:datetime(), calendar:datetime(), calendar:time()) -> atom()).
+schedule(StartTime, EndTime, PublishInterval) ->
+  gen_server:call(?SERVER, {schedule, StartTime, EndTime, PublishInterval}).
+
+-spec(cancel_schedule(reference()) -> no_return()).
+cancel_schedule(Ref) ->
+  gen_server:cast(?SERVER, {cancel_schedule, Ref}).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+-spec init([]) -> {ok, State :: #state{}}.
+init([]) ->
+  process_flag(trap_exit, true),
+  {ok, #state{refs = maps:new()}}.
+
+-spec handle_call({schedule, StartTime :: calendar:datetime(), EndTime :: calendar:datetime(), PublishInterval :: calendar:time()},
+    From :: {pid(), Tag :: term()}, State :: #state{}) ->
+  {reply, reference(), NewState :: #state{}}.
+handle_call({schedule, StartTime, EndTime, PublishInterval}, _From, State) ->
```

Comment:
You can do this because `{schedule, integer(), pid(), list()}` is subtype of `term()` which is expected as the first argument to `Module:handle_call/3` callback function (see [documentation](http://www.erlang.org/doc/man/gen_server.html#Module:handle_call-3)).


Try changing `From :: {pid(), Tag :: term()}` to `From :: {reference(), Tag :: term()}` - you will get an error.

On the other hand, you can freely manipulate specs of the tt_scheduler:schedule/3. For example dialyzer doesn't complain about:
```erlang
-spec(schedule(reference(), calendar:datetime(), calendar:time()) -> atom()).
schedule(StartTime, EndTime, PublishInterval) ->
  gen_server:call(?SERVER, {schedule, StartTime, EndTime, PublishInterval}).
```

It is because there's no call to this function in the code and dialyzer cannot see any broken contract. However, if you add such a call somewhere with the code base:
```erlang
...
schedule({{2015,12,15}, {10,00,00}},
                {{2015,12,15}, {11,00,00}},
                {0,0,20})
...
```
it will complain:
```bash
_build/default/lib/talks_tweeter/src/tt_scheduler.erl
  42: The call tt_scheduler:schedule({{2015, 12, 15}, {10, 0, 0}}, {{2015, 12, 15}, {11, 0, 0}}, {0,0,20}) breaks the contract (reference(), calendar:datetime(), calendar:time()) -> atom()
```